crates/primitives/transactions/src/from_broadcasted_transactions.rs

```diff
@@ -48,135 +60,206 @@ pub enum BroadcastedTransactionConversionError {
+            fee_data_availability_mode,
+            is_query,
+        }) => {
+            let sierra_contract_class = Felt252Wrapper::from(flattened_contract_class.class_hash()).into();
+            let sierra_program_length = flattened_contract_class.sierra_program.len();
+            let abi_length = flattened_contract_class.abi.len();
+
+            let casm_contract_class = flattened_sierra_to_casm_contract_class(flattened_contract_class)
+                .map_err(|_| BroadcastedTransactionConversionError::SierraCompilationFailed)?;
+
+            let tx =
+                starknet_api::transaction::DeclareTransaction::V3(starknet_api::transaction::DeclareTransactionV3 {
+                    resource_bounds: resource_bounds_mapping_conversion(resource_bounds),
+                    tip: Tip(tip),
+                    signature: TransactionSignature(
+                        signature.into_iter().map(|v| Felt252Wrapper::from(v).into()).collect(),
+                    ),
+                    nonce: Felt252Wrapper::from(nonce).into(),
+                    class_hash: sierra_contract_class,
+                    compiled_class_hash: Felt252Wrapper::from(compiled_class_hash).into(),
+                    sender_address: Felt252Wrapper::from(sender_address).into(),
+                    nonce_data_availability_mode: data_availability_mode_conversion(nonce_data_availability_mode),
+                    fee_data_availability_mode: data_availability_mode_conversion(fee_data_availability_mode),
+                    paymaster_data: PaymasterData(
+                        paymaster_data.into_iter().map(|v| Felt252Wrapper::from(v).into()).collect(),
+                    ),
+                    account_deployment_data: AccountDeploymentData(
+                        account_deployment_data.into_iter().map(|v| Felt252Wrapper::from(v).into()).collect(),
+                    ),
+                });
 
-                let contract_class = ContractClass::V1(
-                    ContractClassV1::try_from(casm_contract_class)
-                        .map_err(|_| BroadcastedTransactionConversionError::CasmContractClassConversionFailed)?,
-                );
-
-                UserTransaction::Declare(tx, contract_class)
+            let tx_hash = tx.compute_hash(chain_id, is_query);
+            let contract_class = ContractClass::V1(
+                ContractClassV1::try_from(casm_contract_class)
+                    .map_err(|_| BroadcastedTransactionConversionError::CasmContractClassConversionFailed)?,
+            );
+
+            {
+                let class_info = ClassInfo::new(&contract_class, sierra_program_length, abi_length)?;
+                if is_query {
+                    DeclareTransaction::new_for_query(tx, tx_hash, class_info)
+                } else {
+                    DeclareTransaction::new(tx, tx_hash, class_info)
+                }
+                .map_err(|_| BroadcastedTransactionConversionError::InvalidTransactionVersion)?
```

Comment:
looks repetitive, should we move this to a function

Reply:
It is the only place it is used, so I see no need to factorize it.
Also it is an outcome of a bad interface design by blockfier, so I hope it will be fixed on their side one of those days.
See: https://github.com/starkware-libs/blockifier/issues/1725

Reply:
it's being reused in V3 and V2 no? but it might be too small to refactor?

Reply:
No you are right, it's used 3 times. My bad